packages/flutter/lib/src/material/page_transitions_theme.dart

```diff
@@ -0,0 +1,363 @@
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    return CupertinoPageRoute.buildPageTransitions<T>(route, context, animation, secondaryAnimation, child);
+  }
+}
+
+/// Defines the page transition animations used by [MaterialPageRoute]
+/// for different [TargetPlatform]s.
+///
+/// The [MaterialPageRoute.buildTransitions] method looks up the current
+/// current [PageTransitionsTheme] with `Theme.of(context).pageTransitionsTheme`
+/// and delegates to [buildTransitions].
+///
+/// If a builder with a matching platform is not found, then the
+/// [FadeUpwardsPageTransitionsBuilder] is used.
+///
+/// See also:
+///
+///  * [ThemeData.pageTransitionsTheme], which defines the default page
+///    transitions for the overall theme.
+///  * [FadeUpwardsPageTransitionsBuilder], which defines a default page transition.
+///  * [OpenUpwardsPageTransitionsBuilder], which defines a page transition
+///    that's similar to the one provided by Android P.
+///  * [CupertinoPageTransitionsBuilder], which defines a horizontal page
+///    transition that matches native iOS page transitions.
+@immutable
+class PageTransitionsTheme extends Diagnosticable {
+  /// Construct a PageTransitionsTheme.
+  ///
+  /// By default the list of builders is: [FadeUpwardsPageTransitionsBuilder],
+  /// [CupertinoPageTransitionsBuilder] for [TargetPlatform.android]
+  /// and [TargetPlatform.iOS] respectively.
+  const PageTransitionsTheme({ Map<TargetPlatform, PageTransitionsBuilder> builders }) : _builders = builders;
+
+  static final Map<TargetPlatform, PageTransitionsBuilder> _defaultBuilders = <TargetPlatform, PageTransitionsBuilder>{
+    TargetPlatform.android: const FadeUpwardsPageTransitionsBuilder(),
+    TargetPlatform.iOS: const CupertinoPageTransitionsBuilder(),
+  };
+
+  /// The [PageTransitionsBuilder]s supported by this theme.
+  Map<TargetPlatform, PageTransitionsBuilder> get builders => _builders ?? _defaultBuilders;
+  final Map<TargetPlatform, PageTransitionsBuilder> _builders;
+
+  /// Delegates to the builder for the current [ThemeData.platform]
+  /// or [FadeUpwardsPageTransitionsBuilder].
+  ///
+  /// [MaterialPageRoute.buildTransitions] delegates to this method.
+  Widget buildTransitions<T>(
+    PageRoute<T> route,
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    TargetPlatform platform = Theme.of(context).platform;
+
+    if (CupertinoPageRoute.isPopGestureInProgress(route))
+      platform = TargetPlatform.iOS;
+
+    final PageTransitionsBuilder matchingBuilder =
+      builders[platform] ?? const FadeUpwardsPageTransitionsBuilder();
+    return matchingBuilder.buildTransitions<T>(route, context, animation, secondaryAnimation, child);
+  }
+
+  List<PageTransitionsBuilder> _all(Map<TargetPlatform, PageTransitionsBuilder> builders) {
+    return TargetPlatform.values.map((TargetPlatform platform) => builders[platform]).toList();
```

Comment:
might be good to precache this on creation, since i imagine (hope?) that this class will be compared more often than created...